source/service.go

```diff
@@ -292,9 +292,28 @@ func (sc *serviceSource) extractHeadlessEndpoints(svc *v1.Service, hostname stri
 				headlessDomains = append(headlessDomains, fmt.Sprintf("%s.%s", pod.Spec.Hostname, hostname))
 			}
 
+		NextDomain:
 			for _, headlessDomain := range headlessDomains {
 				var ep string
-				if sc.publishHostIP {
+				if sc.publishHostIP && externalIP {
+					if pod.Status.Phase == v1.PodRunning {
+						node, err := sc.nodeInformer.Lister().Get(pod.Spec.NodeName)
+						if err != nil {
+							log.Errorf("Unable to find node where Pod %s is running: %v", pod.Spec.Hostname, err)
+							return nil
+						}
+						for _, a := range node.Status.Addresses {
+							if a.Type == v1.NodeExternalIP {
+								log.Debugf("Generating matching endpoint %s with NodeExternalIP %s", headlessDomain, a.Address)
+								targetsByHeadlessDomain[headlessDomain] = append(targetsByHeadlessDomain[headlessDomain], a.Address)
+								continue NextDomain
+							}
+						}
+					} else {
+						log.Debugf("Pod %s/%s is not running", pod.GetNamespace(), pod.GetName())
+						return nil
```

Comment:
won't returning here make this method return nothing in case when this runs with pods rolling, not yet ready? Is that desired?

Reply:
Yes but it should re-sync again on another period (default=60s)

Reply:
ah makes sense, cool